File: go/recursion.go
package main

import (
	rl "github.com/gen2brain/raylib-go/raylib"
	"math"
)

const (
	screenFactor int32 = 80
	width        int32 = 16 * screenFactor
	height       int32 = 9 * screenFactor

	branches         int     = 5
	branch_len       int32   = 2 * screenFactor
	branch_thickness float32 = 10.0
	branch_angle     float64 = 2 * math.Pi / float64(branches)
	branch_level     int32   = 6
)

func drawSnowflake(center rl.Vector2, level int32, length int32, thickness float32, hue float32) {
	if level <= 0 {
		return
	}
	color := rl.ColorFromHSV(hue, 1.0, 1.0)
	for i := 0; i <= branches-1; i++ {
		branch := rl.Vector2{
			X: center.X + float32(length)*float32(math.Cos(branch_angle*float64(i))),
			Y: center.Y + float32(length)*float32(math.Sin(branch_angle*float64(i))),
		}
		rl.DrawLineEx(center, branch, thickness, color)
		drawSnowflake(branch, level-1, int32(float32(length)*0.5), thickness*0.5, hue+70)
	}
}

func main() {
	rl.InitWindow(width, height, "Copo de nieve")
	defer rl.CloseWindow()

	rl.SetTargetFPS(60)
	for !rl.WindowShouldClose() {
		rl.BeginDrawing()
		rl.ClearBackground(rl.Black)
		center := rl.Vector2{
			X: float32(rl.GetScreenWidth()) * 0.5,
			Y: float32(rl.GetScreenHeight()) * 0.5,
		}
		drawSnowflake(center, branch_level, branch_len, branch_thickness, 0.0)
		rl.EndDrawing()
	}
}

File: c/recursion.c
#include <math.h>
#include "raylib.h"

// window props
#define SCREEN_FACTOR 80
#define WIDTH 16*SCREEN_FACTOR
#define HEIGHT 9*SCREEN_FACTOR
// snowflake branches
#define BRANCHES 5
#define BRANCH_LEN SCREEN_FACTOR*2
#define BRANCH_THICKNESS 10.0
#define BRANCH_ANGLE 2.0*M_PI/BRANCHES
#define BRANCH_LEVEL 6

// recursive snowflake
void drawSnowflake(Vector2 center, int level, int length, float thick, float hue)
{
    if (level <= 0) return;
    Color color = ColorFromHSV(hue, 1.0, 1.0);
    for (int i = 0; i <= BRANCHES-1; i++) {
        Vector2 branch = {center.x + cos(BRANCH_ANGLE*i)*length, center.y + sin(BRANCH_ANGLE*i)*length};
        DrawLineEx(center, branch, thick, color);
        drawSnowflake(branch, level-1, length*0.5, thick*0.5, hue+70.0);
    }
}

int main(void)
{
    // SetConfigFlags(FLAG_WINDOW_RESIZABLE);
    InitWindow(WIDTH, HEIGHT, "Copo de nieve");
    SetTargetFPS(60);

    while (!WindowShouldClose())
    {
        BeginDrawing();
        ClearBackground(BLACK);
        Vector2 center = {GetScreenWidth()*0.5, GetScreenHeight()*0.5};
        drawSnowflake(center, BRANCH_LEVEL, BRANCH_LEN, BRANCH_THICKNESS, 0.0);
        EndDrawing();
    }
    CloseWindow();

    return 0;
}
